src/formatter.rs

```diff
@@ -93,23 +89,16 @@ fn highlight_code<'a, E>(
 }
 
 /// Print a token stream to an ANSI terminal.
-pub fn print_lines<T>(
-    writer: &mut T,
-    lines: impl Iterator<Item = LineType>,
-    style: &StyleConfig,
+pub fn highlight_lines<L, F, E>(
+    lines: L,
+    yield_snippet: &mut F,
     keep_empty_lines: bool,
-) -> Result<(), TealdeerError>
+) -> Result<(), E>
 where
-    T: Write,
+    L: Iterator<Item = LineType>,
+    F: for<'snip> FnMut(HighlightingSnippet<'snip>) -> Result<(), E>,
```

Comment:
What kind of syntax is `for<'snip>`? 😄 I'm not familiar with that.

Reply:
It means that we want `F` to be generic over the lifetime `'snip`. We actually need it to be able to handle multiple disjoint lifetimes (each item in line iterator clears the buffer, so the references from the previous iteration become invalid)

Reply:
Cool!